src/ocaml/mode.ml

```diff
@@ -135,3 +137,135 @@ module Dict = struct
          { byte; native })
   end
 end
+
+module Select = struct
+  type mode = t
+
+  type nonrec t =
+    | Only of t
+    | All
+
+  let of_option = function
+    | None -> All
+    | Some m -> Only m
+
+  let is_not_all = function
+    | All -> false
+    | Only _ -> true
+
+  let equal t t' =
+    match (t, t') with
+    | Only m, Only m' -> equal m m'
+    | All, All -> true
+    | _, _ -> false
+
+  let hash = Poly.hash
+
+  let to_dyn t =
+    let open Dyn in
+    match t with
+    | All -> Variant ("All", [])
+    | Only m -> Variant ("Only", [ to_dyn m ])
+
+  let encode t =
+    let open Dune_sexp.Encoder in
+    match t with
+    | All -> string "all"
+    | Only m -> encode m
+
+  let decode =
+    let open Dune_sexp.Decoder in
+    let+ value = either (keyword "all") decode in
+    match value with
+    | Left () -> All
+    | Right mode -> Only mode
+end
+
+module MultiDict = struct
+  (* Here we use a Hashtbl to be able in the future to use more variants than
+     just "Byte, Native or All". Since values are always added and never deleted
+     or modified we take advantage of the fact that when using [add t key value]
+     previous bindings for [key] are not removed but simply hidden and that
+     [find_all] returns all these bindings to represent list of values with a
+     hashtbl for single values. *)
+  module Hashtbl = Hashtbl.Make (Select)
+
+  type mode = t
+
+  type 'a t = 'a Hashtbl.t
+
+  let create () = Hashtbl.create 30
```

Comment:
size of key type is closer to 3, but to be fair I'd be suspicious of any magic different from `0` in a call to `Hashtbl.create`

Reply:
obsolete